### DISCUSSION TITLE
Slow response in browser-authorization process

### DISCUSSION THREAD

**Discussion by PeterJBassett** [iTwin/community#54] [Q&A] [answered]
Hello I work on developing the FutureOn connector to iTwin used by Shell for their Bentley Plantsight projects.

I have upgraded my application to use the browser-authorization module for user sign-in.
See https://github.com/iTwin/auth-clients/tree/main/packages/browser#readme

The following problems are occuring:
1.  The response to calling `signInSilent` can be very slow, it often takes about 20 seconds to respond.
2. Sometimes when calling `signInSilent` it doesn't realise that the user is already signed in, so (after the 20 seconds), it pops up the sign in window, and immediately closes it again.

The code I am using is below.

```
    try {
      console.log('Attempting signInSilent')
      await authClient.signInSilent()
    } catch {
      console.log('Performing signIn Popup')
      await authClient.signInPopup()
      // Note: this opens a popup for the user to sign in, and then 
      // will reopen this page with the redirect URL. 
      // The onMount code above is called and invokes the handleSignInCallback
      // which should close the popup and return here.
    } 
```

**Comment by ben-polinsky** [comment]
Hello,

There's a chance this is related to a change made on the IMS backend. I'll take a look at let you know if there's a fix we can make on our end.

**Comment by ben-polinsky** [comment]
I'm not able to reproduce on a test viewer application. Are there any failed or slow network calls? Specifically to `ims.bentley.com`?

**Reply by PeterJBassett** [reply]
Hello @ben-polinsky thank you very much for your reply. 
Looking in the chrome console, I can see the following logs.

It seems to be something to do with handling third-party cookies. It generates 52 of these logs during the 10s it takes to trigger the pop-up. The logs about third-party cookies are something to do with ims.bentley.com queries.

See screenshots below:


<img width="623" alt="Screenshot 2024-10-18 at 09 20 14" src="https://github.com/user-attachments/assets/02b329e7-8b2c-4302-a862-33ea430655f4">



<img width="623" alt="Screenshot 2024-10-18 at 09 20 41" src="https://github.com/user-attachments/assets/7c41829b-29d6-4b4b-a9be-d8a32e236877">

**Reply by PeterJBassett** [reply]
@ben-polinsky in the logs shown above, this is the URL string that is being logged in the 52 logs:
```
https://ims.bentley.com/connect/authorize?
client_id=spa-m19YIiH1EpMzsd9RYBGa9jIJL
&redirect_uri=https%3A%2F%2Fpeter-itwin-integration-frontenduser.futureon-dev.fieldtwin.com%2Fredirect
&response_type=code
&scope=email+openid+profile+organization+itwinjs+itwin-platform
&state=751b398f9ace424f8ca4998a293b1cc2
&code_challenge=4_pIypR_wVDATzkpmF39anvrgTxM4O6rLgXWSxypgGM
&code_challenge_method=S256
&response_mode=query
&prompt=none
```

**Comment by ben-polinsky** [comment]
Hm.. I do see these issues in my browser, yet I am not experiencing the slow responses you are seeing. 

If you check the network tab in chrome devtools, is there a certain request or requests that are taking the 10 seconds? 
You noted you upgraded to the browser authorization package. What were you previously using which did not produce this behavior?

I have a few ideas for trying to narrow down the issue:

1. Could you try a different browser?
2. Could you push your code (or at least the auth part) for me to test? 
3. And/or try out your setup with this react template app: https://github.com/iTwin/viewer/tree/master/packages/templates/cra-template-web-viewer?

**Reply by PeterJBassett** [reply]
Hello, @ben-polinsky thanks again for your help.
1. I have tried using Edge, and get the same delay. I guess because Edge is based on Chrome.
2. See below for the code I am using.
3. I will try out the template app, and let you know.

```
<script>
/**
 * @module SignIn.svelte
 * @description Provides the SignIn component for the user to sign in to iTwin.
 * This follows the process desribed for the @itwin/browser-authorization npm module.
 * See https://github.com/iTwin/auth-clients/tree/main/packages/browser
 * and the OIDC obstacle course see https://developer.bentley.com/apis/overview/authorization/
 */

  import { onMount } from 'svelte'
  import { BrowserAuthorizationClient } from "@itwin/browser-authorization"
  import itwinLogo from '../assets/itwin-logo.svg'

  // The callback function to be called when the user signs in
  let { onSignIn } = $props()

  // Determine the redirect URI needed for the OIDC authorisation flow
  // This must match with the redirect URIs listed for the SPA app
  // registered on https://developer.bentley.com/my-apps/
  const redirectURL = new URL(document.location.href)
  const redirectURI = `${redirectURL.origin}/redirect`

  // The SPA App config for the connector application 'futureon-itwin-connector-1'
  // registered at https://developer.bentley.com/my-apps/
  const authConfig = {
    clientId: 'spa-m19YIiH1EpMzsd9RYBGa9jIJL',
    redirectUri: redirectURI,
    postSignoutRedirectUri: "",
    responseType: "code",
    authority: "https://ims.bentley.com",
    scope: `email openid profile organization itwinjs itwin-platform`,
  }

  // The authorisation client for the OIDC login process
  const authClient = new BrowserAuthorizationClient(authConfig)

  // The component sign-in state
  let signInState = $state({
    awaitingSignIn: false,
    isPopup: false,
    isSignedIn: false,
    infoMsg: '',
    errorMsg: '',
    jwt: '',
  })

  // The sign-in result
  let signInResult = $state({
    jwt: '',
    isSignedIn: false,
    errorMsg: '',
  })

  onMount(async () => {
    // Note that the component is mounted twice. Once to display the initial Sign In page
    // and a second time when the Bentley Sign In popup is invoked by the redirect step of the
    // OIDC sign in process.
  
    signInState.awaitingSignIn = false

    // Indicates if the window is the redirected popup Bentley sign in window 
    signInState.isPopup = (window.opener != null)

    // Indicates if the user has successfully signed-in in the redirected popup window
    signInState.isSignedIn = document.location.href.includes('redirect?code')

    // Indicates if the sign-in failed in the redirected popup window
    signInState.errorMsg = (document.location.href.includes('redirect?error')) ?
      'Sign In Failed' : ''

    // When the user successfully signs-in (in the popup window)
    // invoke the callback to close the popup window and return 
    // to the parent window
    if (signInState.isPopup) {
      if (signInState.isSignedIn || signInState.errorMsg != '') {
        await authClient.handleSigninCallback()
      }
    }
    return true
  });

  /// signIn: button event handler
  async function signIn() {
    signInState.awaitingSignIn = true
    try {
      console.log('Attempting signInSilent')
      await authClient.signInSilent()
    } catch {
      console.log('Performing signIn Popup')
      await authClient.signInPopup()
      // Note: this opens a popup for the user to sign in, and then 
      // will reopen this page with the redirect URL generated Bentley. 
      // The onMount code above is then called and invokes the handleSignInCallback
      // which should close the popup and return here in the parent window.
    } 
    // At this point (back in the parent window) the user has signed in
    // And so should now be able to obtain the JWT
    try {
      signInResult.jwt = await authClient.getAccessToken()
      signInResult.errorMsg = ''
      signInResult.isSignedIn = true
      console.log('Successfully signed in')
    } catch (e) {
      signInResult.errorMsg = 'Bentley Sign In Failed'
      signInResult.jwt = ''
      signInResult.isSignedIn = false
      console.log('Failed to sign in')
    }
    // Call the props callback function to pass the JWT back to the parent component
    onSignIn(signInResult)
    signInState.awaitingSignIn = false

    // Set a timeout to refresh the session
    setSessionRefresh()
    return true
  }

  /// signOut: button event handler (used for testing)
  async function signOut() {
    await authClient.signOut()
  }

  /// Set a timeout to refresh the session
  function setSessionRefresh() {
    setTimeout(async () => {
      console.log('Refreshing the session')
      await signIn()
    }, 25*60*1000)
  }

</script>

```

**Reply by PeterJBassett** [reply]
@ben-polinsky To answer your other question. Previously, I was using my own implementation of the Bentley OATH2 process. It was much more complicated, a number of js files, but it did work, and didn't have the delay. But because I am updating the code to use Svelte 5, I wanted to use the standard itwin npm package to keep the code simpler.

**Reply by PeterJBassett** [reply]
@ben-polinsky Here is some more information about the problem I am having.
* if I am already signed-in to Bentley, then there is no delay, and it quickly returns the Access Token
* If I sign out, and then sign-in again, this is when I get the 20s delay.

The delay __always__ is 20s. This consists of 2 * 10s delays.
* Calling `await authClient.signInSilent()` takes 10s before going to the `catch` block
* In the catch block `await authClient.signInPopup()` takes 10s before the popup window appears.

In the `BrowserAuthorizationClient` code there is a variable `var DefaultSilentRequestTimeoutInSeconds = 10;`
and `silentRequestTimeoutInSeconds = DefaultSilentRequestTimeoutInSeconds,`

So it looks like this timeout could have something to do with it.

**Reply by PeterJBassett** [reply]
@ben-polinsky Here is some more info.
I tried changing my sign-in code as below:
```
  async function signIn() {
    signInState.awaitingSignIn = true
    try {
      console.log('Attempting signInSilent')
      await authClient.getAccessToken()
      // await authClient.signInSilent()
    } catch {
      console.log('Performing signIn Popup')
      await authClient.signInPopup()
      // Note: this opens a popup for the user to sign in, and then 
      // will reopen this page with the redirect URL generated by Bentley. 
      // The onMount code above is then called and invokes the handleSignInCallback
      // which should close the popup and return here in the parent window.
    } 
```
Making this change made reduced the overall delay to 10s, because the `await authClient.getAccessToken()` fails immediately rather than having a timeout. So it immediately goes to the `catch` block.

**Reply by ben-polinsky** [reply]
Thanks for fleshing this out a bit more. I'll see what I can dig up. Where in the world are you located?

**Reply by PeterJBassett** [reply]
Hello @ben-polinsky I am based in the UK.

**Reply by PeterJBassett** [reply]
Also, it might make a difference that I am not an internal Bentley user, but am an external user logging in to Bentley.

**Reply by ben-polinsky** [reply]
Circling back for clarification: have you run the old version of the application _recently_ and validated there's no delay?

**Reply by PeterJBassett** [reply]
Hi @ben-polinsky Yes, the old version has never had this delay.
The old implementation of my application did not use the @itwin/browser-authorization module, and doesn't have this timeout in its code.

**Reply by ben-polinsky** [reply]
Hello again. I've been able to reproduce on a new app built with vite, but can't reproduce using webpack. You're correct that the timeout is the culprit. Our browser client wraps [oidc-client-ts](https://github.com/authts/oidc-client-ts/) and we do not expose the option to customize the silent renew timeout, but I can make that change.

I'm still stumped as to why webpack fails fast and vite hits the timeout...

**Reply by ben-polinsky** [reply]
Update. In the case of the app built with vite, the issue was that the `redirectUrl` and my main app were using the same url. This probably triggered an error of some sort in the hidden iframe, or otherwise prevented the `handleSigninCallback()` call from completing.

Check that the `redirectUrl` used to initialize the client corresponds to a route that only handles the sign in callback and nothing else.  With this code, the `signinSilent()` call toward the bottom fails immediately:

```ts
import { BrowserAuthorizationClient } from '@itwin/browser-authorization'
const redirectUri = "http://localhost:5173/signin-callback"

// ensure the redirect url does nothing else but handle the callback
if (redirectUri === window.location.pathname) {
  BrowserAuthorizationClient.handleSignInCallback().then(() => {
    console.log("handled")
  })
} else {
  // ensure the client isn't being instantiated on the redirectUri route.
  const client = new BrowserAuthorizationClient({
    redirectUri,
    scope: "itwin-platform",
    responseType: "code",
    authority: "https://ims.bentley.com",
    ...
  })

  document.querySelector<HTMLDivElement>('#app')!.innerHTML = `
    <div>
      <a href="https://bentley.com" target="_blank">
        <img src="https://www.bentley.com/wp-content/uploads/2022/04/bentley-logo-400X100-black.svg" class="logo" alt="Vite logo" />
      </a>
      <a href="https://itwinjs.org" target="_blank">
        <img src="https://pbs.twimg.com/profile_images/1050473874058563585/Fs0iAFxh_400x400.jpg" class="logo" alt="Vite logo" />
      </a>
      <h1>iTwin.js Demo App</h1>
      <div class="card">
      ${client.isAuthorized ? 'Signed in' : 'Not signed in'}
        <button id="sign-in" type="button">Sign in</button>
      </div>
    </div>
  `
  document.querySelector("#sign-in")!.addEventListener("click", async () => {
    try {
      await client.getAccessToken() // throws immediately
    } catch (error) {
      await client.signInSilent() // now fails fast and throws
    }
  })
}
```

Whether or not this is the underlying issue you're facing, the fact that you are hitting the timeout does indicate something is wrong with your implementation.

**Reply by PeterJBassett** [reply]
Hello @ben-polinsky thank you very much for your help. 
I will try updating my app as you have suggested.

**Reply by PeterJBassett** [reply]
Hello @ben-polinsky thanks again for all your help.
I have found that if I use `oidc-client-ts` to log in to Bentley everything works instantly, with no delays.

```
  async function signIn() {
    let user
    try {
      user = await userManager.getUser()
      if (!user) {
        await userManager.signinPopup()
        // the popup calls userManager.signinPopupCallback()
        user = await userManager.getUser()
      }
      if (user) {
        console.log('Successfully signed in', user)
        signInResult.jwt = user.auth_token
        ...

```

Unfortunately, I realise that this does not give the right `auth_token`, and that `@itwin/browser-authorization` probably does some extra step to get the actual iModel access token. So this doesn't solve my problem. So I will stick with using  `@itwin/browser-authorization`

The question is, why does there have to be a timeout, why cant it call `oidc-client-ts` `userManager.getUser()` to find out if the user is already signed in.

**Reply by aruniverse** [reply, marked as answer]
You are probably just missing the `token_type` which is just `Bearer`
The access token we expect to be used with iTwin Platform will be `${user.token_type} ${user.access_token}`

**Reply by PeterJBassett** [reply]
Hello @aruniverse - yes of course that was it!!. So it now works using `oidc-client-ts` directly,
Thank you very much for your help.

**Reply by ben-polinsky** [reply]
> I have found that if I use `oidc-client-ts` to log in to Bentley everything works instantly, with no delays.

Curious - which version? :) We're on 2.x. Maybe they've fixed this behavior in 3.x.

**Reply by PeterJBassett** [reply]
Hello @ben-polinsky it is version 3.1.0
